lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Migrations.hs

```diff
@@ -0,0 +1,481 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Test.Integration.Scenario.API.Shelley.Migrations
+    ( spec
+    ) where
+
+import Prelude
+
+import Cardano.Mnemonic
+    ( entropyToMnemonic, genEntropy )
+import Cardano.Wallet.Api.Types
+    ( ApiTransaction
+    , ApiUtxoStatistics
+    , ApiWallet
+    , ApiWalletMigrationInfo (..)
+    , DecodeAddress
+    , EncodeAddress (..)
+    , WalletStyle (..)
+    )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( PaymentAddress )
+import Cardano.Wallet.Primitive.AddressDerivation.Byron
+    ( ByronKey )
+import Cardano.Wallet.Primitive.AddressDerivation.Icarus
+    ( IcarusKey )
+import Cardano.Wallet.Primitive.AddressDerivation.Shelley
+    ( ShelleyKey )
+import Control.Monad
+    ( forM_ )
+import Data.Generics.Internal.VL.Lens
+    ( view, (^.) )
+import Data.Maybe
+    ( mapMaybe )
+import Data.Quantity
+    ( Quantity (..) )
+import Data.Text
+    ( Text )
+import Data.Word
+    ( Word64 )
+import Test.Hspec
+    ( SpecWith, describe, it, shouldBe, shouldSatisfy )
+import Test.Integration.Framework.DSL
+    ( Context (..)
+    , Headers (..)
+    , Payload (..)
+    , emptyIcarusWallet
+    , emptyRandomWallet
+    , emptyWallet
+    , eventually
+    , expectErrorMessage
+    , expectField
+    , expectResponseCode
+    , fixturePassphrase
+    , fixtureWallet
+    , getFromResponse
+    , icarusAddresses
+    , json
+    , listAddresses
+    , randomAddresses
+    , request
+    , unsafeRequest
+    , verify
+    , walletId
+    , (.>)
+    )
+import Test.Integration.Framework.TestData
+    ( errMsg400ParseError
+    , errMsg403NothingToMigrate
+    , errMsg403WrongPass
+    , errMsg404NoWallet
+    )
+
+import qualified Cardano.Wallet.Api.Link as Link
+import qualified Cardano.Wallet.Api.Types as ApiTypes
+import qualified Data.Map.Strict as Map
+import qualified Network.HTTP.Types.Status as HTTP
+
+
+spec :: forall n t.
+    ( DecodeAddress n
+    , EncodeAddress n
+    , PaymentAddress n ShelleyKey
+    , PaymentAddress n IcarusKey
+    , PaymentAddress n ByronKey
+    ) => SpecWith (Context t)
+spec = do
+    it "SHELLEY_CALCULATE_01 - \
```

Comment:
Do you think we could mimic the "new clean way of writing scenarios" that we have in recently written integration spec files? (i.e., declare scenario as separate function and reference them in the main "spec")

Reply:
To be honest I don't find that new way handy for me. I recognize this is cleaner/more aesthetic maybe:

```
    describe "BYRON_MIGRATE" $ do
        scenario_MIGRATE_01 @n fixtureRandomWallet
        scenario_MIGRATE_02 @n fixtureRandomWallet 1
        scenario_MIGRATE_02 @n fixtureRandomWallet 3
        scenario_MIGRATE_02 @n fixtureRandomWallet 10
        scenario_MIGRATE_02 @n fixtureIcarusWallet 1
        scenario_MIGRATE_02 @n fixtureIcarusWallet 3
        scenario_MIGRATE_02 @n fixtureIcarusWallet 10
```
But you have to scroll down/scroll up to see what particular test is doing. This is ok(ish) when there is small number of scenarios in a file, but when there are many of them it becomes painful (to me).

Ok, maybe that's my personal preference (or I just got used to it so much), but I find good ol `spec -> describe -> it` scheme (+ some helper methods for repeatable actions) a good compromise between "clean" and sufficient level of detail in the (integration) test suite such that you can immediately tell what is going on...